File: occasum-core/src/main/java/pw/stamina/occasum/registry/PropertyRegistry.java
package pw.stamina.occasum.registry;

import pw.stamina.occasum.PropertyHandle;
import pw.stamina.occasum.node.PropertyNode;
import pw.stamina.occasum.node.factory.PropertyNodeFactory;
import pw.stamina.occasum.properties.Property;

import java.util.Map;
import java.util.Optional;

public interface PropertyRegistry {

    PropertyNode register(PropertyHandle handle, Property property);

    void registerAll(PropertyHandle handle, Iterable<Property> properties);

    void unregisterAll(PropertyHandle handle);

    void unregister(PropertyNode node);

    boolean unregister(PropertyHandle handle, String path);

    Optional<PropertyNode> findRootNode(PropertyHandle handle);

    PropertyNode findOrCreateRootNode(PropertyHandle handle);

    //TODO: Implement
    //PropertyNode findNodeOrCreateFolder(PropertyHandle handle, String path);

    Map<PropertyHandle, PropertyNode> findAllRootNodes();

    static PropertyRegistry standard(PropertyNodeFactory factory) {
        return new StandardPropertyRegistry(factory);
    }

    static PropertyRegistry decorateNullRejecting(PropertyRegistry registry) {
        return new NullRejectingPropertyRegistryDecorator(registry);
    }
}

File: occasum-core/src/main/kotlin/pw/stamina/occasum/registry/PropertyRegistry.kt
package pw.stamina.occasum.registry

import pw.stamina.occasum.PropertyHandle
import pw.stamina.occasum.node.PropertyNode
import pw.stamina.occasum.node.factory.PropertyNodeFactory
import pw.stamina.occasum.properties.Property

interface PropertyRegistry {

    fun register(handle: PropertyHandle, property: Property<*>): PropertyNode

    fun registerAll(handle: PropertyHandle, properties: Iterable<Property<*>>)

    fun unregisterAll(handle: PropertyHandle)

    fun unregister(node: PropertyNode)

    fun unregister(handle: PropertyHandle, path: String): Boolean

    fun findRootNode(handle: PropertyHandle): PropertyNode?

    fun findOrCreateRootNode(handle: PropertyHandle): PropertyNode

    //TODO: Implement
    //PropertyNode findNodeOrCreateFolder(PropertyHandle handle, String path);

    fun findAllRootNodes(): Map<PropertyHandle, PropertyNode>

    companion object {

        fun standard(factory: PropertyNodeFactory): PropertyRegistry {
            return StandardPropertyRegistry(factory)
        }
    }
}
